icicle/src/runtime.cpp

```diff
@@ -274,9 +274,28 @@ extern "C" eIcicleError icicle_load_backend(const char* path, bool is_recursive)
   };
 
   auto load_library = [](const char* filePath) {
-    ICICLE_LOG_DEBUG << "Attempting load: " << filePath;
-    void* handle = dlopen(filePath, RTLD_LAZY | RTLD_GLOBAL);
-    if (!handle) { ICICLE_LOG_ERROR << "Failed to load " << filePath << ": " << dlerror(); }
+    // Convert the file path to a std::string for easier manipulation
+    std::string path(filePath);
+
+    // Extract the library name from the full path
+    std::string fileName = path.substr(path.find_last_of("/\\") + 1);
+
+    // Check if the library name contains "icicle" and if the path contains "/backend/"
+    if (fileName.find("icicle") == std::string::npos || path.find("/backend/") == std::string::npos) {
+      ICICLE_LOG_DEBUG << "Skipping: " << filePath << " - Not an Icicle backend library.";
+      return;
+    }
+
+    // Check if the library name contains "device". If yes, load it with GLOBAL visibility, otherwise LOCAL.
+    // The logic behind it is to avoid symbol conflicts by using LOCAL visibility but allow backends to expose symbols
+    // to the other backend libs. For example to reuse some device context or any initialization required by APIs that
+    // we want to do once.
```

Comment:
Shouldn't all backends be independent of each other? When would we want backends to rely on other ones?

Reply:
it's not that backend depend on each other. When you build a backend such as cuda, then you would typically have a backend lib per frontend lib so that you only use the field/curves you need. So you will have  one device library that implement the device API interface and then you will have backend lib per field/curve. It may the case that the backend needs some global state. in CUDA case it's the license and it will probably be in others. Other than that in other APIs you sometimes have some initialization of objects at runtime that you maybe need to share so that's why I want to expose those symbols. The field/curve libs typically don't need to expose any symbol and we had a bug in CUDA curve that for some reason it mixed symbols (Seems like nvcc bug hoenstly but nevermind).